pkg/addonmanager/controllers/agentdeploy/hookcontroller.go

```diff
@@ -0,0 +1,306 @@
+			if _, ok := c.agentAddons[accessor.GetName()]; !ok {
+				return false
+			}
+
+			return true
+		},
+		addonInformers.Informer()).
+		WithFilteredEventsInformersQueueKeyFunc(
+			func(obj runtime.Object) string {
+				accessor, _ := meta.Accessor(obj)
+				return fmt.Sprintf("%s/%s", accessor.GetNamespace(), accessor.GetLabels()[constants.AddonLabel])
+			},
+			func(obj interface{}) bool {
+				accessor, _ := meta.Accessor(obj)
+				if accessor.GetLabels() == nil {
+					return false
+				}
+
+				addonName, ok := accessor.GetLabels()[constants.AddonLabel]
+				if !ok {
+					return false
+				}
+
+				if _, ok := c.agentAddons[addonName]; !ok {
+					return false
+				}
+
+				if accessor.GetName() != hookWorkName(addonName) {
+					return false
+				}
+				return true
+			},
+			workInformers.Informer(),
+		).
+		WithSync(c.sync).ToController("addon-hook-deploy-controller", recorder)
+}
+
+func (c *addonHookDeployController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	key := syncCtx.QueueKey()
+	klog.V(4).Infof("Reconciling addon hook deploy %q", key)
+
+	clusterName, addonName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// ignore addon whose key is not in format: namespace/name
+		return nil
+	}
+
+	agentAddon, ok := c.agentAddons[addonName]
+	if !ok {
+		return nil
+	}
+
+	// Get ManagedCluster
+	managedCluster, err := c.managedClusterLister.Get(clusterName)
+	if errors.IsNotFound(err) {
+		c.cache.removeCache(hookWorkName(addonName), clusterName)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if !managedCluster.DeletionTimestamp.IsZero() {
+		// managed cluster is deleting, do nothing
+		return nil
+	}
+
+	managedClusterAddon, err := c.managedClusterAddonLister.ManagedClusterAddOns(clusterName).Get(addonName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			c.cache.removeCache(hookWorkName(addonName), clusterName)
+			return nil
+		}
+		return err
+	}
+	managedClusterAddonCopy := managedClusterAddon.DeepCopy()
+	owner := metav1.NewControllerRef(managedClusterAddon, addonapiv1alpha1.GroupVersion.WithKind("ManagedClusterAddOn"))
+
+	objects, err := agentAddon.Manifests(managedCluster, managedClusterAddon)
+	if err != nil {
+		return err
+	}
+	if len(objects) == 0 {
+		return nil
+	}
+	_, hookWork, err := buildManifestWorkFromObject(clusterName, addonName, objects)
+	if err != nil {
+		return err
+	}
+
+	if hookWork != nil {
+		if !hasFinalizer(managedClusterAddonCopy.GetFinalizers(), constants.PreDeleteHookFinalizer) {
+			managedClusterAddonCopy.Finalizers = append(managedClusterAddonCopy.Finalizers, constants.PreDeleteHookFinalizer)
+			_, err = c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).Update(
+				ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+			return err
+		}
+	}
+
+	if managedClusterAddon.DeletionTimestamp.IsZero() {
+		return nil
+	}
+	// apply hookWork when addon is deleting
+	if hookWork != nil {
```

Comment:
I think you can do 

```
if hookWork == nil {
  return nil 
}
```
here?

Reply:
I think you should alwasy removeFinalizer when hookWork == nil

Reply:
done